src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -22,26 +22,72 @@
 --   This optimization does not work when functions are applied only partially.
 --   Thus, we have to convert partial applications to full applications.
 --
+--   We differentiate between top-level partial applications (i.e. when a 
+--   function is defined as the partial application of another defined function 
+--   or constructor) and partial applications that occur in the arguments of 
+--   the function declaration's right-hand side. 
+-- 
+--   We perform regular η-conversions on partial applications in proper 
+--   subexpressions of a function declaration's right-hand side. 
+-- 
+--   However, on the top-level, we add the missing arguments to the left-hand and
+--   right-hand sides of the function rule explicitly, without a lambda abstraction. 
+--   This is an optimization that allows the compiler to avoid some unnecessary 
+--   monadic lifting.
+--
 --   = Specification
 --
 --   == Preconditions
 --
 --   The arity of all constructors and functions must be known (i.e., there
---   must be corresponding environment entries).
+--   must be corresponding environment entries) and all function declarations 
+--   must be type annotated.
 --
 --   == Translation
 --
---   On the right-hand sides of function declarations, all of the largest
+--   Assume that we have the following function declaration. 
+--
+--   > f e₁ … eₖ = g @α₁ … @αₚ e₁ … eₘ
+--
+--   where @g@ is the name of an @n@-ary constructor or function declaration and 
+--   @m < n@. This declaration is then replaced by 
+--   > f e₁ … eₖ x₍ₘ₊₁₎ … xₙ = g @α₁ … @αₚ e₁ … eₘ x₍ₘ₊₁₎ … xₙ
```

Comment:
Some markup languages have problems with consecutive block-lvel elements. I don't know how Haddock handles this case but I'd insert a blank line between block-level elements just to be safe.

```suggestion
--
--   > f e₁ … eₖ x₍ₘ₊₁₎ … xₙ = g @α₁ … @αₚ e₁ … eₘ x₍ₘ₊₁₎ … xₙ
```